server/src/main/java/org/opensearch/cluster/metadata/Metadata.java

```diff
@@ -175,6 +175,27 @@ public enum XContentContext {
     public interface Custom extends NamedDiffable<Custom>, ToXContentFragment, ClusterState.FeatureAware {
 
         EnumSet<XContentContext> context();
+
+        static Custom fromXContent(XContentParser parser, String name) throws IOException {
+            try {
+                return parser.namedObject(Custom.class, name, null);
+            } catch (NamedObjectNotFoundException e) {
+                logger.warn("Unknown custom object with type {}", name);
+                parser.skipChildren();
```

Comment:
This fromXContent is called when we are parsing the custom metadata object or when we are parsing the full metadata.
In case, we are parsing a single custom metadata object and we face an unknown object, there is no need to call `skipChildren()` as we are going to skip parsing the entire object.
Otherwise, when we are parsing the whole metadata object, we need to skip the children so that we can move on to parse the next object in the metadata.

I think your point is valid, that `skipChildren()` should be caller's responsibility only.